pkg/gateway/model_build_targetgroup.go

```diff
@@ -172,65 +164,82 @@ func (t *latticeServiceModelBuildTask) buildTargets(ctx context.Context) error {
+		tg, err = t.buildTargetGroupForServiceExportCreation(ctx, tgName)
+	} else {
+		// build TargetGroup Model for serviceExport deletion
+		tg, err = t.buildTargetGroupForServiceExportDeletion(ctx, tgName)
+	}
+	if err != nil {
+		return err
+	}
+	t.tgByResID[tgName] = tg
+	t.targetGroup = tg
+	return nil
+}
 
+func (t *targetGroupModelBuildTask) buildTargetGroupForServiceExportCreation(ctx context.Context, targetGroupName string) (*latticemodel.TargetGroup, error) {
 	svc := &corev1.Service{}
 	if err := t.Client.Get(ctx, k8s.NamespacedName(t.serviceExport), svc); err != nil {
 		glog.V(6).Infof("Error finding corresponding service %v error :%v \n", k8s.NamespacedName(t.serviceExport), err)
 		// mark there is no serviceexport dependence on the TG
-		t.Datastore.SetTargetGroupByServiceExport(tgName, false, false)
-		return err
+		t.Datastore.SetTargetGroupByServiceExport(targetGroupName, false, false)
+		return nil, err
 	}
-
-	//if t.serviceExport.
-	tgSpec := latticemodel.TargetGroupSpec{
-		Name: tgName,
+	tg := latticemodel.NewTargetGroup(t.stack, targetGroupName, latticemodel.TargetGroupSpec{
+		Name: targetGroupName,
 		Type: latticemodel.TargetGroupTypeIP,
 		Config: latticemodel.TargetGroupConfig{
 			VpcID: config.VpcID,
 			// Fill in default HTTP port as we are using target port anyway.
-			Port:                80,
-			IsServiceImport:     false,
-			IsServiceExport:     true,
-			K8SServiceName:      t.serviceExport.Name,
-			K8SServiceNamespace: t.serviceExport.Namespace,
-			Protocol:            "HTTP",
-			ProtocolVersion:     vpclattice.TargetGroupProtocolVersionHttp1,
+			Port:                  80,
+			IsServiceImport:       false,
+			IsServiceExport:       true,
+			K8SServiceName:        t.serviceExport.Name,
+			K8SServiceNamespace:   t.serviceExport.Namespace,
+			K8SHTTPRouteName:      "", // When build TG model for serviceExport, the httproutename should be ""
+			K8SHTTPRouteNamespace: "",
+			Protocol:              "HTTP",
+			ProtocolVersion:       vpclattice.TargetGroupProtocolVersionHttp1,
 		},
-	}
-
-	tg := latticemodel.NewTargetGroup(t.stack, tgName, tgSpec)
-	glog.V(6).Infof("buildTargetGroup, tg[%s], tgSpec%v \n", tgName, tg)
-
-	// add targetgroup to localcache for service reconcile to reference
-	// for serviceexport, the httproutename is set to ""
-	t.Datastore.AddTargetGroup(tgName, "", "", "", tgSpec.Config.IsServiceImport, "")
-
-	if !t.serviceExport.DeletionTimestamp.IsZero() {
-		// triggered by serviceexport delete
-		t.Datastore.SetTargetGroupByServiceExport(tgName, false, false)
-	} else {
-		// triggered by serviceexport add
-		t.Datastore.SetTargetGroupByServiceExport(tgName, false, true)
-	}
+	})
+	glog.V(6).Infof("buildTargetGroup, tg[%s], tgSpec%v \n", targetGroupName, tg)
+	// add targetgroup to Datastore local cache for service reconcile to reference
+	// for serviceexport, the httproutename is ""
+	t.Datastore.AddTargetGroup(targetGroupName, "", "", "", tg.Spec.Config.IsServiceImport, "")
```

Comment:
we need to refactor datastore that it does not except wrong arguments, especially when we expect some of them being empty strings only

Reply:
Talk with @solmonk , current datastore indeed hard to understand and need refactoring, but we should do it after GRPCRoute and TargetGroupPolicy delivered.  


And @solmonk also suggest to fix code diff as small as possible

What I am think is:
For Scenario 1: stand-alone serviceExport deletion leak tg, that can be fix by few lines change, but I am not 100% sure need to double confirm)
And For scenario 2: non-stand-alone serviceExport deletion leak tg (non-stand-alone means the serviceImport of that serviceExport referenced in a httproute), I think it need to change where we add/update/delete datastore targetGroup cache, only refactoring datastore could make that code easy to follow.

My next step would be try to use a seperate PR to fix Scenario 1 only(keep code diff as small as possible), and keep current PR open until we delivered the GRPCRoute and TargetGroupPolicy 

If quick fix for Scenario 1, I will just go ahead to work on something else